tools/windows-node-installer/pkg/cloudprovider/aws/ec2.go

```diff
@@ -606,53 +641,87 @@ func (a *AwsProvider) createWindowsWorkerSg(infraID string, vpc *ec2.Vpc) (strin
 	return *sg.GroupId, nil
 }
 
-// authorizeSgIngress attaches inbound rules for RDP from user's external IP address and all traffic within the VPC
-// on newly created security group for Windows instance. The function returns error if failed.
-func (a *AwsProvider) authorizeSgIngress(sgID string, vpc *ec2.Vpc) error {
-	myIP, err := GetMyIp()
-	if err != nil {
-		return err
+// GetPortsFoSg returns a slice of ports which need to be updated in the rules for local IP
+func getPortsForSg(myIP string, rules []*ec2.IpPermission) []int64 {
+	var ports []int64
+	portTracker := map[int64]bool{}
+	// Loop through the rules, then loop through the ips to get a map of ports
+	if rules != nil {
+		for _, rule := range rules {
+			for _, ips := range rule.IpRanges {
+				if *ips.CidrIp == myIP+"/32" {
+					portTracker[*rule.FromPort] = true
+				}
+			}
+		}
 	}
 
-	_, err = a.EC2.AuthorizeSecurityGroupIngress(&ec2.AuthorizeSecurityGroupIngressInput{
-		GroupId: aws.String(sgID),
-		IpPermissions: []*ec2.IpPermission{
+	// Add ports that need to be updated in a slice ports
+	if !portTracker[WINRM_PORT] {
+		ports = append(ports, WINRM_PORT)
+	}
+	if !portTracker[sshPort] {
+		ports = append(ports, sshPort)
+	}
+	if !portTracker[rdpPort] {
+		ports = append(ports, rdpPort)
+	}
+	return ports
+}
+
+// GetRulesForSgUpdate returns a slice of rules which are required to be updated in sg with local IP
+// This serves as an input to addIngressRules function. This function returns error if failed.
+func (a *AwsProvider) GetRulesForSgUpdate(myIP string, rules []*ec2.IpPermission, vpcCidr string) ([]*ec2.IpPermission, error) {
+	var hasClusterCidrRule bool
+	rulesForUpdate := make([]*ec2.IpPermission, 0)
+
+	ports := getPortsForSg(myIP, rules)
+
+	if rules != nil {
+		// check if cluster communication rule is missing
+		for _, rule := range rules {
+			for _, ips := range rule.IpRanges {
+				if *ips.CidrIp == vpcCidr && *rule.IpProtocol == "-1" {
+					hasClusterCidrRule = true
+					break
```

Comment:
We don't loop over the rules multiple times. Please move this into `getPortsForSg()` and make this return a bool in addition to the list. The function name will need to be more generic.